doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -192,43 +197,110 @@ structures. To get around this issue, we can introduce internal action IDs to en
 ```c++
 std::map<std::wstring, unsigned int> _NameMap;
 std::map<KeyChord, unsigned int> _KeyMap;
-std::map<unsigned int, Action> _IDMap;
 
-unsigned int _nextID = 1;
+// This could be stored as a `map<unsigned int, Action>`,
+//  but a vector provides this functionality much better.
+std::vector<Action> _ActionList;
 ```
 
 With this design, the internal action ID can be saved as an `unsigned int`. This ID is never exposed.
- When an `Action` is added to the `ActionMap`, we update `_IDMap` and increment `_nextID`.
+ When an `Action` is added to the `ActionMap`, we update `_ActionList`.
 
 Now that there are no duplicates between `_NameMap` and `_KeyMap`, creating a copy of `_NameMap` and
- `_KeyMap` is trivial. We can iterate over `_IDMap` to create a copy of every `Action` stored
- and save it to a new `_IDMap`.
+ `_KeyMap` is trivial. We can iterate over `_ActionList` to create a copy of every `Action` stored
+ and save it to a new `_ActionList`.
+
+### Layering Actions
+
+We need a way to determine where an action came from to minimize how many actions we serialize when we
+ write to disk. This is a two part approach that happens as we're loading the settings
+ 1. Load defaults.json
+     - For each of the actions in the JSON...
+         - Construct the `Command` (basically the `Command::LayerJson` we have today)
+         - Add it to the `ActionMap`
+           - this should update `_NameMap`, `_KeyMap`, and `_ActionList` appropriately
+           - if the newly added name/key chord conflicts with a pre-existing one,
+              redirect `_NameMap` or `_KeyMap` to the newly added `Command` instead,
+              and update the conflicting one.
+2. Load settings.json
+     - Add a parent to `ActionMap`
+         - The purpose of a parent is to continue a search when the current `ActionMap`
+            can't find a `Command` for a query. The parent is intended to be immutable.
+     - Load the actions array like normal (see step 1)
+
+Introducing a parent mechanism to `ActionMap` allows it to understand where a `Command`
+ came from. This allows us to minimize the number of actions we serialize when we write
+ to disk, as opposed to serializing the entire list of actions.
+
+`ActionMap` queries will need to check their parent when they cannot find a matching `Command`
+ in their `_ActionList` (or sooner).
+
+### Modifying Actions
+
+There are several ways a command can be modified:
+- change/remove the key chord
+- change the name
+- change the icon
+- change the action

```

Comment:
If we change the action, is it really the same action at all anymore? 🤔

Reply:
Good point. Changing the action is going to be the biggest pain because it'll change the `InternalActionID`. We'll definitely have to regenerate any views here.